Classes/common/HTTP/CDTURLSessionInterceptorContext.h

```diff
@@ -0,0 +1,39 @@
+//
+//  CDTURLSEssionFilterContext.h
+//  
+//
+//  Created by Rhys Short on 17/08/2015.
+//
+//
+
+#import <Foundation/Foundation.h>
+#import "CDTMacros.h"
+
+@interface CDTURLSessionInterceptorContext : NSObject
+
+@property (nonnull, readonly, nonatomic, strong) NSMutableURLRequest *request;
+
+- (nullable instancetype) init UNAVAILABLE_ATTRIBUTE;
+
+/**
+ *  Initalizes a CDTURLSessionInterceptorContext
+ *
+ *  @param request the request this context should represent
+ *
+ **/
+- (nullable instancetype) initWithRequest:(nonnull NSMutableURLRequest*)request NS_DESIGNATED_INITIALIZER;
+
+
+@end
+
+@protocol CDTURLSessionInterceptor <NSObject>
+
+/**
+ *  Intercepts a request before it is made
+ *  @param context the context for this interception
+ * 
+ *  @return the context for this interception
+ **/
+@optional - (nonnull CDTURLSessionInterceptorContext*) interceptRequestInContext:(nonnull CDTURLSessionInterceptorContext*)context;
+
+@end
```

Comment:
`@optional` moved to its own line in 646b643cb312cd5479411d1da9755bce7012968d 
I'm not sure if the protocol should be in it's own header or not, I think we should a discussion about that